compiler/next/lib/types/Type.cpp

```diff
@@ -19,13 +19,83 @@
 
 #include "chpl/types/Type.h"
 
+#include "chpl/types/AnyType.h"
+#include "chpl/types/BoolType.h"
+#include "chpl/types/BuiltinType.h"
+#include "chpl/types/ComplexType.h"
+#include "chpl/types/ImagType.h"
+#include "chpl/types/IntType.h"
+#include "chpl/types/PrimitiveType.h"
+#include "chpl/types/RealType.h"
+#include "chpl/types/UintType.h"
+#include "chpl/types/UnknownType.h"
+
 namespace chpl {
 namespace types {
 
 
 Type::~Type() {
 }
 
+static void gatherType(Context* context,
+                       std::unordered_map<UniqueString,const Type*>& map,
+                       const char* c_str,
+                       const Type* t) {
+  auto name = UniqueString::build(context, c_str);
+  map.insert( {name, t} );
+}
+
+static
+void gatherPrimitiveType(Context* context,
+                         std::unordered_map<UniqueString,const Type*>& map,
+                         const PrimitiveType* p) {
+  gatherType(context, map, p->c_str(), p);
+}
+
+void Type::gatherBuiltins(Context* context,
+                          std::unordered_map<UniqueString,const Type*>& map) {
+
+  gatherPrimitiveType(context, map, BoolType::get(context, 0));
```

Comment:
It bothers me a little bit that you have to hardcode a list of primitive types and their applicable bit widths in here.

Maybe it's not the sort of thing likely to change: but if someone were to make a new primitive type or add a new bit width they'd be responsible for updating this code and it seems likely this would be forgotten.

That said, I'm not sure how I would re-engineer things to avoid this problem.

Reply:
Yeah, some compilers use a separate file (more like TypeClassesList.h) to handle this sort of thing. And in fact we do that for the macro-supported BuiltinTypes.

One thing we could do is to put this gathering process in to PrimitiveType and have that call something in BoolType, IntType, etc. But that seem like over-engineering it to me.

As to the potential problem of adding a new bit width and having to update this and perhaps forgetting to do so - I would expect that if this isn't updated then either it doesn't ever need to be updated or basically nothing with that new bit width will work. So I would expect a developer to figure it out when updating it.

Also, for now I have added all of these integer types but it isn't necessarily strictly necessary. When we parse `int(64)` it is parsed as a call of `int` with argument `64`. When we resolve that we can call `IntType::get(64)`. However it made sense to me to add these different widths since they are "built-in" and I can imagine wanting to run some action on "all the builtin types".